HousingResidentInformationAPI.Tests/V1/UseCase/ValidatePostcodeTests.cs

```diff
@@ -22,6 +22,11 @@ public void ValidPostcodesReturnTrue(string postcode)
         }
 
         [TestCase("1")]
+        [TestCase("111111")]
+        [TestCase("111 111")]
+        [TestCase("EEEEEE")]
```

Comment:
What is in invalid case??

Reply:
The regex pattern is explicitly case insensitive, duplication of the tests for lower case characters will produce identical results. I also didn't cover the examples you have given in mixed case, but it's explicit in the regex pattern that these would also pass or fail, depending on format.
The fact that the regex pattern is already explicitly case insensitive makes these tests more about postcode formats rather than the case of the characters used for input.  

The accompanying test to this has a postcode with a mix of lower and uppercase characters. 

The main factor here is to cover as many common incorrectly formatted inputs as possible.
I also don't have knowledge on all possible current and future postcode formats for the UK. 
Although not perfect/exhaustive I would prefer to stick with this regex pattern, I presume it comes from a source that has tested it well before documenting it for public use. 
Source being: https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/488478/Bulk_Data_Transfer_-_additional_validation_valid_from_12_November_2015.pdf.
The specification for the outcode patterns can be read in the document linked above.
The accompanying error message received by the user when the postcode validation fails also explicitly refers to the postcode format not the character case. 

Does your calculation include missing tests for mixed case postcodes or just lowercase postcodes?
Apart from the lowercase patterns are there any other patterns you think should be added?

Reply:
Yes, I realize it's explicitly case insensitive. My argument is not whether it will work or not work - I know it will work.

What I'm actually saying is that we're supposed to be following Test Driven Development as per our playbook, where tests come first. If you were to TDD this feature you currently have, case-insensitivity should come out of the Test you have written.

Now, I realize that what was done here was not done in TDD, hence I'm saying it's all still fine so long you add the edge case tests that you would have written if you were following the TDD. _(Which would include lowercase tests)_

Reply:
When it comes to adding the lowercase tests, you shouldn't need to go as granular as mixed case. 1 lowercase test per outcode combination should suffice to prove that the pattern works with lowercase. **So that would be 6 tests in total - not that much.**

Reply:
I **agree** with you on that this Regex will suffice for our usecase - that's why I'm not advocating for adding all those exceptional UK postcode formats. :)

All I'm saying is that this pattern contains an explicit GIR 0AA match, which could be removed or left alone (_but if it's left I would want a test)_ And I'm bringing up that this pattern does not contain whitespace support as I'm not aware whether you need it or not,... as this information was not in the Pull Request description _(which should ideally contain a list of requirements you followed in implementing a feature)_

Reply:
It's perfectly possible for the regex pattern to not match mixed case input. How is it possible to infer that the code works for mixed case input without checking the regex pattern, but not infer that the regex pattern works for lowercase input? 
There is also a testcase in the code which uses mixed case input that passes.

I'm still not sure what benefit there is in having an explicit test for gir oaa or GIR OAA or for removing it from the regex. You state that having a test for mixed postcodes is overly granular, I think it also applies here.
The code being cleaner if it's removed is a subjective observation.
There is no G or g postcode in hackney. 
We might as well have a regex pattern that is specifically for hackney postcodes.

Reply:
It's true, you could write a regex in a really horrid and inefficient way, where the test of doing all lowercase postcode wouldn't cover for mixed-case cases.

What I'm saying is more of, when your typical user enters a postcode, it'll most likely be in either all capital, or all lowercase _(if they didn't care enough to press capslock)_.

As a result of such constraint, whenever you're doing TDD, you'd at bare minimum include tests the most common postcode formats in lowercase and uppercase. Currently you only got that for the uppercase, so I'd say add about 6 more test cases and you should be good.

Reply:
It's possible for user input to be in any form, why make an assumption on what a "typical user" input would be over an internet connection. Your assumptions aren't based on any objectively measured observations. Your implied constraint fits the particular case you think is the most common. 
I'm not sure any of this is progressive as you have ignored the chance to answer any of the questions which would be constructive to the discussion. I've responded to your statement in regards to the granularity of test coverage for this postcode format validation.
It's explicit within the regex pattern that the postcode formatting isn't affected by the case of the characters. There is also a test as I've pointed out which uses a mixed case of characters.
This validation is explicitly a format validation of postcodes.

Reply:
I get an impression that the only progressive response on this PR would be: "Looks good to me!" followed by approval.

**I don't know which part of:**
1. telling you which tests I think would be good to add, 
2. listing the cases, 
3. showing examples, 
4. pointing out bugs 
5. and in some cases even giving you fixed regular expressions was not constructive.

I understand what you're saying.
> "The regex pattern is explicitly case insensitive, duplication of the tests for lower case characters will produce identical results."

You're saying: "My implementation explicitly adds lowercase cases, so why test them when it's going to pass just as the uppercase tests?"
Same argument could be said about all the code in this API. You could say:

`"well I copied this part of the gateway from the other API, it's added in this file explicitly. So why test it the second time on this API, when it was tested on the other one? These tests will only give duplicate results."`

So why even have any tests at all? Most our APIs follow the same pattern, most of the code is identical except for variable name differences. Why write any tests again, when it was tested on other APIs - it clearly works there, so the tests here are just going to give duplicate results, right?

Reply:
Your quoted statement can't be applied to this case where a postcode format validation is being tested.
Which bug? the regex pattern itself? It's not a perfect pattern. Does it need to be? 
Your fix is identical to the current pattern except for removing matching on GIR 0AA or gir 0aa.